exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -1663,35 +1634,189 @@ func (b *Bitfinex) ConfigureWS() error {
 	})
 }
 
-// Unsubscribe sends a websocket message to stop receiving data from the channel
-func (b *Bitfinex) Unsubscribe(channelsToUnsubscribe []stream.ChannelSubscription) error {
+// Subscribe sends a websocket message to receive data from channels
+func (b *Bitfinex) Subscribe(channels []stream.ChannelSubscription) error {
+	return b.parallelChanOp(channels, b.subscribeToChan)
+}
+
+// Unsubscribe sends a websocket message to stop receiving data from channels
+func (b *Bitfinex) Unsubscribe(channels []stream.ChannelSubscription) error {
+	return b.parallelChanOp(channels, b.unsubscribeFromChan)
+}
+
+// parallelChanOp performs a single method call in parallel across streams and waits to return any errors
+func (b *Bitfinex) parallelChanOp(channels []stream.ChannelSubscription, m func(*stream.ChannelSubscription) error) error {
+	wg := sync.WaitGroup{}
+	wg.Add(len(channels))
+	errC := make(chan error, len(channels))
+
+	for i := range channels {
+		go func(c *stream.ChannelSubscription) {
+			defer wg.Done()
+			if err := m(c); err != nil {
+				errC <- err
+			}
+		}(&channels[i])
+	}
+
+	wg.Wait()
+	close(errC)
+
 	var errs error
-	for i := range channelsToUnsubscribe {
-		idAny, ok := channelsToUnsubscribe[i].Params["chanId"]
-		if !ok {
-			errs = common.AppendError(errs, fmt.Errorf("cannot unsubscribe from a channel without an id"))
-			continue
-		}
-		chanID, ok := idAny.(int)
-		if !ok {
-			errs = common.AppendError(errs, fmt.Errorf("chanId is not an int"))
-			continue
-		}
+	for err := range errC {
+		errs = common.AppendError(errs, err)
+	}
 
-		req := map[string]interface{}{
-			"event":  "unsubscribe",
-			"chanId": chanID,
+	return errs
+}
+
+// subscribeToChan handles a single subscription and parses the result
+// on success it adds the subscription to the websocket
+func (b *Bitfinex) subscribeToChan(c *stream.ChannelSubscription) error {
+	req, err := subscribeReq(c)
+	if err != nil {
+		return fmt.Errorf("%w: %w; Channel: %s Pair: %s", stream.ErrSubscriptionFailure, err, c.Channel, c.Currency)
+	}
+
+	// Although docs only mention this for wsBook, it works for all chans
+	subID := strconv.Itoa(int(b.Websocket.Conn.GenerateMessageID(false)))
```

Comment:
I _will_ make this requested change, not least to ensure Itoa doesn't get copied to less safe examples.

But:
I thought that GenerateMessageID was safe within 32-bit boundaries?
I even checked before `int()` ing the output that int64 -> int32 would be okay.

Just want to make sure that I understand.

Reply:
The equivalent of `Itoa` is `FormatInt`, not `ParseInt`.
Were you thinking of `Atoi` when you read this?
Not sure if that had a bearing on your comment and why we saw the risk differently here?

Reply:
Yeah I got them confused sorry 😄 😭

Reply:
No concern really, but I'm going to change it anyway to avoid setting a bear trap 🐻 🪤  precedent that less experienced devs might copy.
Specifically: Going `int64` ➡️ `int` (implicit int32 on arch 32) ➡️  `int64` is lossy.
In this contained example it's fine, but since `FormatInt` doesn't return an error we don't lose anything by avoiding it.